src/findex_graph/mod.rs

```diff
@@ -0,0 +1,294 @@
+    type Key: Sized + ZeroizeOnDrop + Sync + Send;
+
+    /// Error type returned by the GxEnc scheme.
+    type Error: std::error::Error + Sync + Send;
+
+    /// Generates a new random seed.
+    fn gen_seed(&self, rng: &mut impl CryptoRngCore) -> Self::Seed;
+
+    /// Deterministically derives a key from the given seed.
+    fn derive_keys(&self, seed: &Self::Seed) -> Self::Key;
+
+    /// Queries the encrypted graph for the given tags and returns the
+    /// decrypted values.
+    async fn get<
+        Tag: Debug + Hash + Eq + Clone + AsRef<[u8]> + From<Vec<u8>>,
+        Value: Hash + Eq + Clone + From<Vec<u8>>,
+        F: Future<Output = bool>,
+        Interrupt: Fn(HashMap<Tag, HashSet<IndexedValue<Tag, Value>>>) -> F,
+    >(
+        &self,
+        key: &Self::Key,
+        tags: HashSet<Tag>,
+        label: &Label,
+        interrupt: &Interrupt,
+    ) -> Result<HashMap<Tag, HashSet<IndexedValue<Tag, Value>>>, Self::Error>;
+
+    /// Encrypts and inserts the given items into the graph. Returns the set of
+    /// tags added to the index.
+    #[allow(clippy::type_complexity)]
+    async fn insert<Tag: Hash + Eq + AsRef<[u8]>, Value: AsRef<[u8]>>(
+        &mut self,
+        rng: Arc<Mutex<impl CryptoRngCore>>,
+        key: &Self::Key,
+        items: HashMap<Tag, Vec<(Operation, IndexedValue<Tag, Value>)>>,
+        label: &Label,
+    ) -> Result<HashSet<Tag>, Self::Error>;
+}
+
+#[derive(Debug)]
+pub struct FindexGraph<
+    UserError: CallbackErrorTrait,
+    EntryTable: DxEnc<ENTRY_LENGTH, Error = Error<UserError>>,
+    ChainTable: DxEnc<LINK_LENGTH, Error = Error<UserError>>,
+> {
+    pub findex_mm: FindexMultiMap<UserError, EntryTable, ChainTable>,
+}
+
+impl<
+    UserError: CallbackErrorTrait,
+    EntryTable: DxEnc<ENTRY_LENGTH, Error = Error<UserError>>,
+    ChainTable: DxEnc<LINK_LENGTH, Error = Error<UserError>>,
+> FindexGraph<UserError, EntryTable, ChainTable>
+{
+    pub fn new(entry_table: EntryTable, chain_table: ChainTable) -> Self {
+        Self {
+            findex_mm: FindexMultiMap {
+                entry_table,
+                chain_table,
+            },
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        collections::{HashMap, HashSet},
+        fmt::Debug,
+        hash::Hash,
+        sync::{Arc, Mutex},
+    };
+
+    use cosmian_crypto_core::{reexport::rand_core::SeedableRng, CsRng};
+
+    use crate::{
+        edx::in_memory::InMemoryEdx,
+        findex_graph::{FindexGraph, GxEnc, IndexedValue},
+        findex_mm::Operation,
+        ChainTable, DxEnc, EntryTable, Label,
+    };
+
+    async fn user_interrupt<
+        Tag: Debug + Hash + Eq + Clone + AsRef<[u8]> + From<Vec<u8>>,
+        Value: Hash + Eq + Clone + From<Vec<u8>>,
+    >(
+        _res: HashMap<Tag, HashSet<IndexedValue<Tag, Value>>>,
+    ) -> bool {
+        false
+    }
+
+    #[actix_rt::test]
+    async fn test_insert_get() {
+        let rng = Arc::new(Mutex::new(CsRng::from_entropy()));
```

Comment:
I think I recall a discussion we had on chacha 12 rounds for csprng purposes, however I can't remember if we agreed on it being enough rounds.

Reply:
With source https://eprint.iacr.org/2019/1492 saying an attack against chacha7 takes roughly 2^238 time complexity, we can safely say that 12 rounds is a great deal between efficiency and security.

We should still note that it doesn't follow any standard.